File: python/floquet_decoherence_example.py
import numpy as np
import matplotlib.pyplot as plt

# Total Hamiltonian:       H = H0 + Hint*cos(omega*t)
# Bare Hamilotnian:        H0 = 0.5*delta*sigma_z + epsilon*sigma_x 
# Interaction Hamiltonian: Hint = 0.5*Vstr*sigma_z
# Decoherence (dephasing): L = Gammaz * sigma_z

# Define the absorption as the probability of measuring in the state measvec
measvec = np.asarray([1, 0])

# Example values of the Hamiltonian parameters
# energy
epsilon = 0.2
# detuning range
delta_range = np.arange(-6.0, 6.02, 0.02)   
# driving frequency
omega = 1.5;       
# number of photons                        
nphotons = 13                              

# Dephasing operator and rate
Lop = np.asarray([[1,0],[0,-1]])
Gammaz_range = np.asarray([0.005, 0.05, 0.1])

# operator associated with measvec
Aop = np.outer(measvec, measvec)

# Preallocate storage vectors for the absorption spectra and the eigenspectrum
Absorp_av = np.zeros((len(delta_range),len(Gammaz_range)))
spec = np.zeros((2, len(delta_range)))

# loop over dephasing rates to generate spectra
for gc, Gammaz in enumerate(Gammaz_range):  
    # loop over interactions
    for jc, delta in enumerate(delta_range):

        # get the list of Lindblad operators
        Ls = [np.sqrt(Gammaz)*Lop]
        
        # define internal Hamiltonian
        H0 = np.asarray([[delta/2,epsilon],[epsilon,-delta/2]]).astype(complex)
        [evs, evecs] = np.linalg.eigh(np.real(H0)) 
        
        # append spectrum as a function of delta
        spec[:,jc] = evs 
        
        # Use ground state as initial state.
        Rho0 = np.outer(evecs[1], evecs[1])
        # vectorise state
        RhoVec0 = np.reshape(Rho0, (4, 1))
        
        # Interaction part of the Hamiltonian
        Hint = np.asarray([[1,0j],[0j,-1]])/2
        
        # Build up superoperators
        Sys1 = np.eye(len(H0))

        # Bare superoperator (with dephasing operator)
        P0 = Liouvillian(H0, Ls)
        
        # Interaction superoperator (time dependent)
        Pint = Liouvillian(Hint, [])
        
        # timestep
        times = [10/Gammaz]

        # Floquet propagator
        USum = floquet_decoherence(times, omega, nphotons, P0, Pint, average=0)
    
        # Compute propagator in Floquet span and then sum over photon states, 
        # outputting the final summation for density matrix at time times
        Rhot = np.reshape(USum[:,:,0] @ RhoVec0, (2, 2), order='F')
        # Output the probability of measurement results of operator Aop.
        Absorp_av[jc,gc] = np.real(np.trace(Aop @ Rhot))

# plot
fig, ax = plt.subplots(2,1, figsize=(6,4))

# plot spectrum (ground and excited states) as a function of delta
ax[0].plot(delta_range/omega, spec[0,:], 'b', label = r'excited')
ax[0].plot(delta_range/omega, spec[1,:], 'r', label = r'ground')
for mult in np.arange(-3,3.1,1):
    ax[0].axvline(x=mult*omega/omega, ymin=-4, ymax=4, linestyle='dashed',linewidth=1, color = 'black', alpha = 0.5)
ax[0].set_ylim([-4,4])
ax[0].set_ylabel('Energy')
ax[0].legend();
ax[0].set_ylabel(r'Energy ($\lambda$)', fontsize =10);
# plot absorption probability as a function of delta
# over the different dephasing rates
for i in range(3):
    ax[1].plot(delta_range/omega, Absorp_av[:,i], label = r'$\Gamma_z$ = '+ str(Gammaz_range[i]))
ax[1].legend(frameon=False)
ax[1].set_ylim([0,1])
ax[1].set_xlabel(r'detuning  ($\delta/\omega$)', fontsize =10);
ax[1].set_ylabel('Absorption probability', fontsize =10)
for mult in np.arange(-3,3.1,1):
    ax[1].axvline(x=mult*omega/omega, ymin=0, ymax=1, linestyle='dashed',linewidth=1, color = 'black', alpha = 0.5)
fig.savefig('figures/floquet_decoherence.pdf');

File: julia/floquet_decoherence_example.jl
using LinearAlgebra
include("src/functions.jl")
include("src/liouvillian.jl")
include("floquet_decoherence.jl")

# H = H0 + H_int * cos(ω * t)  Total Hamiltonian
# H0 = Δ * σz / 2 + ϵ * σx  Bare Hamiltonian
# H_int = V * σz / 2  Interaction Hamiltonian
# L = Γz * σz  Decoherence (dephasing)

# Define the absorption as the probability of measuring in the excited state
meas_vec = [1, 0]

# Example values of the Hamiltonian parameters
ϵ = 0.2             # energy
Δs = -6.0:0.02:6.02 # detuning range
ω = 1.5             # driving frequency
n_ph = 13           # number of photons

# Dephasing operator and rate
L = [1 0; 0 -1]
Γzs = [0.005, 0.05, 0.1]
A_op = meas_vec * meas_vec'

absorb_av = zeros(length(Δs), length(Γzs))
spectrum = zeros(2, length(Δs))

for (i, Γz) in enumerate(Γzs)
    Ls = [sqrt(Γz) * L]

    for (j, Δ) in enumerate(Δs)
        H_0 = Δ/2 * L + ϵ * [0 1; 1 0]
        vals, vecs = eigen(H_0)
        spectrum[:, j] = vals

        ρ0 = vecs[:, 1] * vecs[:, 1]' # start in ground state
        ρ0_vec = reshape(ρ0, :, 1)

        H_int = L / 2  # interaction Hamiltonian
        system_1 = I(length(H_0))
        P0 = liouvillian(H_int, [])

        times = [10 / Γz]

        U_sum = floquet_decoherence(times, ω, n_ph, P0, P_int)
        ρ_t = reshape(U_sum[:, :, 1] * ρ0_vec, 2, 2)
        absorb_av[j, i] = real(tr(A_op * ρ_t))
    end
end

fig = Figure()
ax1 = Axis(fig[1, 1],
    ylabel = "Energy")
lines!(Δs / ω, spectrum[1, :])
lines!(Δs / ω, spectrum[2, :])

ax2 = Axis(fig[2, 1])
for i in eachindex(Γzs)
    lines!(Δs / ω, -absorb_av[:, i], label = "Γz = $(Γzs[i])")
end

fig